README.md
# WebPA - An Online Peer Assessment Tool for Higher Education

WebPA is an online peer assessment system, or more specifically, a peer-moderated marking system. It is designed for 
teams of students doing groupwork, the outcome of which earns an overall group mark. Each student in a group grades 
their team-mates and their own performance. This grading is then used with the overall group mark to provide each 
student with an individual grade. The individual grade reflects the students contribution to the group.

## Requirements

### PHP
The following versions of PHP are supported for the latest version of WebPA:

* PHP 7.4
* PHP 8.0

Your PHP instance must also have the following extensions enabled:

* MySQLi
* Sessions
* XML

### NPM
Node package manager is required to install feather-icons, an icon package used by WebPA. Once you have downloaded WebPA
navigate to its root on the command line and run:

- `npm install`
- `npm run build`

The first command will retrieve the feather-icons package and the second will move the relevant files to WebPA's _js_ 
directory for use in the application.

## Installation

### Download

The easiest way to download WebPA is with the [composer package manager](https://getcomposer.org) using the following
command:

```
composer create-project --prefer-dist --no-dev webpa/webpa webpa
```

Alternatively you can download the latest release from this repository's [release page](https://github.com/WebPA/WebPA/releases).

### Configuration

WebPA has a number of configuration options allowing you to set your database credentials, SMTP mail host details and 
various other options.

The application comes bundled with a `.env.example` file which lists all of the configuation key-value pairs you can
set. 

For speedy development, you can copy this `.env.example` file to a file called `.env` and change the values to suit your
environment. The path of this file can be set in the `includes/inc_global.php` file.

For production environments, please *avoid* using the `.env` file as storing sensitive credentials in a file could be a 
security risk. Instead you should set these key-pairs as environment variables. In Apache, you can set these in your 
`.htaccess` file as follows:

```
SetEnv DB_HOST localhost
```

At a minimum, you should set the following environmental variables to let WebPA function:

* APP_WWW - URL to your instance of WebPA (set without a closing '/')
* DOC_ROOT - Directory path to the WebPA files (set with a closing '/')
* DB_HOST - Database host
* DB_USER - Database username
* DB_PASS - Database password
* DB_NAME - Database name
* DB_PREFIX - Database table prefix. Usually set to 'pa2_'

For more information on the dotenv file please visit the 
[dotenv package's repository](https://github.com/vlucas/phpdotenv). For more information on setting environmental 
variables in Apache, please [visit Apache's website](https://httpd.apache.org/docs/2.4/mod/mod_env.html#setenv).

### Initialise the Database
     
Run the following scripts to initialise the database (edit the files to change the names and password as required):

- install/webpa2\_database.sql: create the database schema and user account;
- install/webpa2\_tables.sql: create the database tables;
- install/webpa2\_administrator.sql: create an administrator account and sample module.

#### Upgrading an Existing Installation

If you already have WebPA installed and are upgrading from version 3.1.0 or below, please run:

- install/webpa\_security\_update.sql
     
### Login to WebPA

- navigate to root of WebPA application
- enter a username of _admin_ and a password of _admin_
- change the password to something more secure after logging in
		 
Delete the _install_ folder when you're finished.

## LTI Extension

The standard installation of WebPA does not include [LTI](https://www.imsglobal.org/activity/learning-tools-interoperability) (Learning Tools Interoperability) support which allows it to integrate seamlessly with most popular Virtual Learning Environments. This can be added via an extension created by Stephen P Vickers. Please visit [Stephen's site](http://www.spvsoftwareproducts.com/php/webpa-lti/) for instructions on how to obtain and install this extension. 

## Documentation

Documentation for WebPA can be found on the [WebPA project site](http://webpaproject.lboro.ac.uk/).

## Changelog

Please see our [changelog](https://github.com/WebPA/WebPA/blob/master/CHANGELOG.md) for a list of updates for this system.

This project uses [semantic versioning](https://semver.org/) from version 3.0.0 onwards.

## Contributing

We always welcome contributors to WebPA. If you can help with development, testing, or documentation, please submit a pull request to this repository.

## Support

Bugs and feature requests are tracked on this project's [GitHub issue tracker](https://github.com/WebPA/WebPA/issues).

## License

This software is distributed under the [GNU General Public License version 3](https://www.gnu.org/licenses/gpl-3.0.en.html).

You may copy, distribute and modify the software as long as you track changes/dates in source files. Any modifications 
to or software including (via compiler) GPL-licensed code must also be made available under the GPL along with build & 
install instructions.

## Credits

WebPA was originally developed by the Centre for Engineering and Design Education at [Loughborough University](http://www.lboro.ac.uk/) with financial support from [JISC](https://www.jisc.ac.uk/)'s e-Learning Capital Programme.

It continues to be maintained by a number of [open source contributors](https://github.com/WebPA/WebPA/graphs/contributors). We thank them for their time and effort supporting this system.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed
- The icon pack, feather-icons, was loaded via a Content Delivery Network. If 
the network was not available, no icons would display in WebPA. This change 
makes the icon pack load from a local installation instead (PR #107)

### Fixed
- When deleting a user, the process tries to delete all the forms linked to that user although there is no relationship 
between the user and the form in the database. This PR fixes that. (PR #113)

## [3.3.0-RC1] - 2023-01-04
### Fixed
- The URL for the comments report was hardcoded instead of being retrieved from the application's settings. This has now
been resolved (PR #100)

## [3.3.0-RC] - 2022-10-26
### Added
- Justification comments from students about why they gave scores can now be viewed by students after being moderated by academic staff (PR #97, #99)

## [3.2.2] - 2022-10-17
### Fixed
- Split 100 forms can now be imported into WebPA. Previously, only likert scale forms could be imported (PR #98)
- XML Validation errors now display when uploading an invalid form (PR #98)

## [3.2.1] - 2021-12-10
### Fixed
- The environment variable for the database port number now applies to all database connections in WebPA (PR #91)

## [3.2.0] - 2021-12-06
### Added
- You can now specify the MySQL port number in the environment variables (PR #88)

### Fixed
- Fixed bug where students were told they had not submitted to a closed assessment when they had (PR #89)

## [3.1.2] - 2021-04-22
### Changed
- Emails are now triggered via a single script, `jobs/Email.php` instead of calling`tutors/asessments/email/ClosingReminder.php` and `tutors/assessments/email/TriggerReminder.php` directly. 

### Removed
- Removed LDAP functionality and options as the implementation did not work.

### Security
- Fixed a large amount of SQL injection attacks
- Change password hashing to use the native password_hash() function in PHP instead of MD5 hashing which is insecure

## [3.1.1] - 2020-10-02
### Fixed
- Remove LTIAuthenticator class as it is a duplicate of DBAuthenticator as should not be used (PR #77)
- Fix bugs found using static analysis (PR #78)

## [3.1.0] - 2020-09-16
### Changed
- Removed most include and require statements and replaced with PSR-4 autoloading (PR #69)
- Remove all global variables from the application to make it easier to maintain the code (PR #70)

### Fixed
- Change default authenticator to be database instead of SAML (PR #64)

## [3.0.7] - 2020-01-20
### Fixed
- Remove a blank line at the top of a PHP class that was causing a fatal error
- Fixed the display academic drop down which was not displaying past years (PR #62)

## [3.0.6] - 2019-11-18
### Fixed
- Fixed a PHP syntax issue in the class_engcis.php file where a missing closing bracket was causing a fatal error to be thrown. (PR #60) 

## [3.0.5] - 2019-11-08
### Fixed
- When using the LDAP integration, users logging in had their firstname, surname, and email address set to a blank value. This was due to code being added to the system that was specific to one insitution, rather than a generalised implementation. It has been fixed by reverting the LDAP authenticator to the same one as in version 2 (PR #59)

## [3.0.4] - 2019-10-18
### Fixed
- When viewing assessments in a module, if no assessment has been created the dropdown will display the years 1969/70. This has now been fixed to display the current academic year instead (PR #58)

## [3.0.3] - 2019-03-21
### Fixed
- Fix undefined offset notice in the class_dao.php file (PR #48)
- Fix respondent list for assessments not showing (PR #48)
- Fix login issue where activity timestamp was not being stored in the database (PR #48)
- Fix issue preventing users from creating assessments via the assessment wizard (PR #48)

## [3.0.2] - 2018-11-20
### Fixed
- Fixed a bug where the database connection was not being properly closed down, meaning subsequent database calls would always fail, even if the `$DB->open()` function was called as the old, closed connection would not be replaced or reinstantiated (PR #43)
- Fixed a bug where only one year value can be displayed in the assessments and metrics academic year drop down (PR #42)

## [3.0.1] - 2018-11-07
### Fixed
- Fix a bug where a unix timestamps were used in the email notifying users that an assessment had been reopened (PR #32)
- Fix a bug where the resource from the create_xml_parser() object was being assigned by reference, causing a strict notice to be issued by PHP (PR #39)
- Fix a bug where accessing undefined array index in the class_dao fetch_value() function caused a warning to be issued (PR #37)
- Fix a bug where tutors who are also students can only see a student menu when logged into WebPA (PR #41)

### Security
- Prevent SQL injection attack for the login and password reset page (PR #40)

## [3.0.0] - 2018-08-06
### Added
- Support for PHP 7.x

## [2.0.0.11] - 2013-12-05

## [1.1.0.1] - 2008-07-19

[Unreleased]: https://github.com/WebPA/WebPA/compare/v3.3.0-RC1...HEAD
[3.3.0-RC1]: https://github.com/WebPA/WebPA/compare/v3.3.0-RC...v3.3.0-RC1
[3.3.0-RC]: https://github.com/WebPA/WebPA/compare/v3.2.2...v3.3.0-RC
[3.2.2]: https://github.com/WebPA/WebPA/compare/v3.2.1...v3.2.2
[3.2.1]: https://github.com/WebPA/WebPA/compare/v3.2.0...v3.2.1
[3.2.0]: https://github.com/WebPA/WebPA/compare/v3.1.2...v3.2.0
[3.1.2]: https://github.com/WebPA/WebPA/compare/v3.1.1...v3.1.2
[3.1.1]: https://github.com/WebPA/WebPA/compare/v3.1.0...v3.1.1
[3.1.0]: https://github.com/WebPA/WebPA/compare/v3.0.7...v3.1.0
[3.0.7]: https://github.com/WebPA/WebPA/compare/v3.0.6...v3.0.7
[3.0.6]: https://github.com/WebPA/WebPA/compare/v3.0.5...v3.0.6
[3.0.5]: https://github.com/WebPA/WebPA/compare/v3.0.4...v3.0.5

src/login.php
<?php
/**
 * Login page
 *
 * Added a link to /accounts/reset.php ('Forgotten your password?')
 * made by Morgan Harris [morgan@snowproject.net] as of 15/10/09
 *
 * @copyright Loughborough University
 * @license https://www.gnu.org/licenses/gpl-3.0.en.html GPL version 3
 *
 * @link https://github.com/webpa/webpa
 */

require_once 'includes/inc_global.php';

use WebPA\includes\functions\Common;

// --------------------------------------------------------------------------------
// Process GET/POST

$msg = Common::fetch_GET('msg', null);

switch ($msg) {
  case 'connfailed':
        $message_class = 'warning';
        $message = 'A connection to the authentication server could not be established.<br />Please try again later.';
        break;
  // --------------------
  case 'denied':
        $message_class = 'warning';
        $message = 'You attempted to access a restricted page.<br />It may be that your session has timed out so please re-enter your details.';
        break;
  // --------------------
  case 'no access':
        $message_class = 'warning';
        $message = 'Your account has been disabled.<br />Please contact support if you do not think this should be the case.';
        break;
  // --------------------
  case 'invalid':
        $message_class = 'warning';
        $message = 'Your username and password were rejected.<br />Please check your details and try again.';
        break;
  // --------------------
  case 'cookies':
        $message_class = 'warning';
        $message = 'Unable to connect to ' . APP__NAME . '; please ensure that your browser is not blocking third-party cookies';
        break;
  // --------------------
  case 'logout':
        $message_class = 'info';
        $message = 'You have logged out.<br />If you wish to log back in, please re-enter your details.';
        break;
  // --------------------
  default:
        $message_class = 'info';
        $message = 'To start using ' . APP__NAME . ' you have to log in.';
        break;
}

// --------------------------------------------------------------------------------
// Begin Page

$UI->page_title = APP__NAME . ' Login';
$UI->menu_selected = '';
$UI->help_link = '?q=node/26';
$UI->breadcrumbs = [
  'login page'  => null,
];


$UI->head();
?>
<style type="text/css">
<!--

p.warning { color: #f00; }
p.info { color: #000; }

-->
</style>
<script language="JavaScript" type="text/javascript">
<!--

  var username_focussed = false;
  var password_focussed = false;

  function username_focus() {
    if ( (!username_focussed) && (!password_focussed) ) { document.getElementById('username').focus(); }
  }

//-->
</script>
<?php
$UI->body('onload="username_focus();"');
$UI->content_start();
?>


<?php echo "<p class=\"$message_class\">$message</p>"; ?>

<div class="content_box">

  <p>Please enter your details below:</p>

  <form action="login_check.php" method="post" name="login_form" style="margin-bottom: 2em;">
  <div style="width: 300px;">
    <table class="form" cellpadding="2" cellspacing="1" width="100%">
    <tr>
      <th><label for="username">Username</label></th>
      <td><input type="text" name="username" id="username" maxlength="30" size="10" value="" onfocus="username_focussed=true" onblur="username_focussed=false" /></td>
    </tr>
    <tr>
      <th><label for="password">Password</label></th>
      <td><input type="password" name="password" id="password" maxlength="16" size="10" value="" onfocus="password_focussed=true" onblur="password_focussed=false" /></td>
    </tr>
    </table>

    <div class="form_button_bar">
      <input class="safe_button" type="submit" name="submit" value="login" />
    </div>
  </div>
  </form>
  <p><strong><a href="accounts/reset.php">Forgotten your password?</a></strong></p>
  <p>This site requires cookies - If you have trouble logging in, please check your cookie settings.</p>

</div>

<?php
$UI->content_end(false);
?>
